cryodrgn/analysis.py

```diff
@@ -92,21 +111,28 @@ def cluster_kmeans(z, K, on_data=True, reorder=True):
     labels = kmeans.fit_predict(z)
     centers = kmeans.cluster_centers_
 
+    centers_ind = None
     if on_data:
         centers, centers_ind = get_nearest_point(z, centers)
 
     if reorder:
         g = sns.clustermap(centers)
         reordered = g.dendrogram_row.reordered_ind
         centers = centers[reordered]
-        if on_data:
+        if centers_ind is not None:
             centers_ind = centers_ind[reordered]
         tmp = {k: i for i, k in enumerate(reordered)}
         labels = np.array([tmp[k] for k in labels])
     return labels, centers
 
 
-def cluster_gmm(z, K, on_data=True, random_state=None, **kwargs):
+def cluster_gmm(
+    z,
+    K: int,
+    on_data: bool = True,
+    random_state: Union[int, np.random.RandomState, None] = None,
```

Comment:
There was a similar comment left in PR 187 too, which I admit I don't get. This type is nasty (to the extent that it is) because that's what the implementation provides (and different code paths do make use of these union types, otherwise I wouldn't have been forced to annotate it this way). Simply being explicit about the types (which is the only correct thing to do here) doesn't make it any more nastier than it already was.

If the issue here is that you want all these elaborate types specified in a single place - `cryodrgn.types` would be the obvious choice, then that's reasonable and I agree with it (though in this case I doubt this typedef is used in other places, if at all). If you're saying this code should only have accepted `int` (as an example), then that's a code refactoring issue that is independent of this PR.

Reply:
I'm fine with having these Union types in the analysis.py functions. Maybe we should be more careful in the core libraries.